File: src/Zahy.Settlement/Zahy.Settlement.Domain/UsageRecord.cs
using System;
using Volo.Abp;
using Volo.Abp.Domain.Entities.Auditing;

namespace Zahy.Settlement;

/// <summary>
/// U1 — a metered-usage record: how many UNITS a merchant consumed for a given partner in a given
/// period (e.g. Pizza House used 6,200 messages via WhatsApp Co in 2026-06). Keyed to
/// partner + merchant + period. APPEND / ACCUMULATE model — usage can be recorded incrementally and
/// summed per period (see <see cref="UsageReports.UsageForPeriod"/>), so several rows for the same
/// partner+merchant+period are expected and add up.
///
/// PURE DATA — holding this row computes NO fee and posts NO journal. Usage-based pricing/posting is
/// the separately-gated U3 phase. The trial balance is never affected by usage records.
/// </summary>
public class UsageRecord : FullAuditedAggregateRoot<Guid>
{
    public Guid PartnerId { get; private set; }

    /// <summary>The merchant (tenant) that consumed the units.</summary>
    public Guid MerchantId { get; private set; }

    public int PeriodYear { get; private set; }

    public int PeriodMonth { get; private set; }

    /// <summary>The consumed unit, e.g. "messages". Free-form label — no pricing is attached here.</summary>
    public string UnitLabel { get; private set; } = SettlementUsageConsts.DefaultUnitLabel;

    /// <summary>Quantity consumed in this (incremental) record. Non-negative; rows accumulate per period.</summary>
    public decimal Quantity { get; private set; }

    /// <summary>Provenance: "seed" / "manual" / "partner-feed" (the gated auto-feed seam).</summary>
    public string Source { get; private set; } = SettlementUsageConsts.DefaultSource;

    /// <summary>The period as a value (derived from the scalar Year/Month columns, never stored).</summary>
    public SettlementPeriod Period => SettlementPeriod.Of(PeriodYear, PeriodMonth);

    protected UsageRecord()
    {
    }

    public UsageRecord(
        Guid id,
        Guid partnerId,
        Guid merchantId,
        SettlementPeriod period,
        string unitLabel,
        decimal quantity,
        string? source = null)
        : base(id)
    {
        if (partnerId == Guid.Empty)
        {
            throw new AbpException("Usage record requires a partnerId.");
        }

        if (merchantId == Guid.Empty)
        {
            throw new AbpException("Usage record requires a merchantId.");
        }

        Check.NotNull(period, nameof(period));
        if (quantity < 0m)
        {
            throw new AbpException($"Usage quantity cannot be negative ('{quantity}').");
        }

        PartnerId = partnerId;
        MerchantId = merchantId;
        PeriodYear = period.Year;
        PeriodMonth = period.Month;
        UnitLabel = NormalizeUnit(unitLabel);
        Quantity = quantity;
        Source = NormalizeSource(source);
    }

    private static string NormalizeUnit(string unitLabel)
    {
        if (string.IsNullOrWhiteSpace(unitLabel))
        {
            return SettlementUsageConsts.DefaultUnitLabel;
        }

        return Check.Length(unitLabel.Trim(), nameof(unitLabel), SettlementUsageConsts.MaxUnitLabelLength);
    }

    private static string NormalizeSource(string? source)
    {
        if (string.IsNullOrWhiteSpace(source))
        {
            return SettlementUsageConsts.DefaultSource;
        }

        return Check.Length(source.Trim(), nameof(source), SettlementUsageConsts.MaxSourceLength);
    }
}

File: frontend/partner-portal-web/src/lib/usage/usageRecord.ts
/**
 * U1 — usage-metering model (frontend mock, mirrors the backend `UsageRecord`). Records how many UNITS
 * a merchant consumed for a partner in a period (e.g. message count) as a QUANTITY. PURE DATA: no fee is
 * computed and no journal is posted here — usage-based pricing/posting is the separately-gated U3 phase.
 */

/** A single (incremental) usage record. Keyed to partner + merchant + period; rows accumulate. */
export interface UsageRecord {
  id: string;
  partnerId: string;
  /** The merchant (tenant) that consumed the units. */
  tenantId: string;
  /** Period as YYYY-MM (mirrors the backend SettlementPeriod.ToString()). */
  period: string;
  /** The consumed unit, e.g. "messages". Free-form label — no pricing attached. */
  unitLabel: string;
  /** Quantity consumed in this record (non-negative). */
  quantity: number;
  /** Provenance: "seed" / "manual" / "partner-feed" (the gated auto-feed seam). */
  source: string;
}

/** Form/append input — the next code/id is assigned by the store. */
export interface UsageRecordInput {
  partnerId: string;
  tenantId: string;
  period: string;
  unitLabel?: string;
  quantity: number;
  source?: string;
}

export const DEFAULT_UNIT_LABEL = "units";
export const DEFAULT_USAGE_SOURCE = "manual";

/**
 * AUTO-FEED — SEAM ONLY (not wired). A real partner/gateway usage feed WOULD push usage records through
 * here. Mirrors the Track B gateway-route seam: the flag stays false and the push path throws, proving
 * no live feed silently ingests usage. Do NOT implement a real feed here (that is the gated phase).
 */
export const USAGE_FEED_WIRED = false;

/** NOT IMPLEMENTED — throws to keep the seam honest (the gated phase wires a real feed). */
export function usageFeedAutoPush(_signals: UsageRecordInput[]): never {
  throw new Error(
    "Usage auto-feed is a SEAM ONLY and is not wired. A real partner/gateway usage feed is enabled " +
      "in the separately-gated phase. Use seeded or manual usage records.",
  );
}
